src/main/java/org/candlepin/resource/CloudRegistrationResource.java

```diff
@@ -160,11 +161,6 @@ else if (version == 2) {
             String errmsg = this.i18n.tr("Cloud registration is not supported by this Candlepin instance");
             throw new NotImplementedException(errmsg, e);
         }
-        catch (CloudRegistrationNotSupportedForOfferingException e) {
-            String errmsg = this.i18n.tr("Cloud registration is not supported for the type of " +
-                "offering the client is using");
-            throw new NotImplementedException(errmsg, e);
-        }
```

Comment:
I  see. Thinking about it a little more, I agree that we should keep the  `CloudRegistrationNotSupportedForOfferingException`, but I don't think that is what should be returned in the case of a non-3P offering type and no mapping in cloud access db. I think that a `NotAuthorizedException` is better for this case, because the problem is that they don't have an association in the cloud access db and not with the cloud offering. I have a condition in `CloudRegistrationResource.validateCloudAuthenticationResult` that checks for this case in the CloudAuthenticationResult.

I think we need CloudRegistrationNotSupportedForOfferingException for a cloud offerings we do not support at all. An example is getting some unsupported cloud offering type that is not 1P, 3P, gold, or custom. I added this as a spec test. In making this change, I also updated `HostedTestApi` and `HostedTestResource` to require a cloud offering type. We were assuming that a null cloud offering type was considered a 3P type, but I think it makes sense to explicitly define the cloud offering types and so we can detect an unknown cloud offering type in `HostedTestCloudRegistrationAdapter`.

I added these changes in a new commit. Let me know what you think!

Reply:
> I think that a `NotAuthorizedException` is better for this case, because the problem is that they don't have an association in the cloud access db and not with the cloud offering

Either of the 2 exceptions could match depending on the context. For example, we plan to support anon org creation for 1P offerings in the future, so the `NotImplementedException` matches there. However for custom offerings, we never plan to support that, so the `NotAuthorizedException` fits more (because in that case they are expected to have created the trust themselves, manually). However, since not supporting1P is a short-term thing,so  `NotAuthorizedException` is the better solution.

> I think we need CloudRegistrationNotSupportedForOfferingException for a cloud offerings we do not support at all. An example is getting some unsupported cloud offering type that is not 1P, 3P, gold, or custom.

There isn't any other type of offering than those, so we don't need an exception for that. Also, RHSM-API/Adapters already have codified throwing `CloudRegistrationNotSupportedForOfferingException` means `we don't have a mapping in the cloud access db, and it is a non-3P offering`, so we can't just change its meaning on our side.

Reply:
I was curious as to how `shouldReturnNotAuthorizedWithRegOnlyOfferTypesAndNoOwnerForV2Auth` is even passing, and it is passing for the wrong reason: in case of no mapping and the offering being non-3P, we should be receiving a `CloudRegistrationNotSupportedForOfferingException`, while what happens now is the adapters return null owner key with isRegistrationOnly=true, and we throw our own exception when we validate that, which isn't what will happen realistically

Reply:
Updated to use `CloudRegistrationNotSupportedForOfferingException` and `NotImplementedException` in the case of cloud account not being mapped to an org and is a type that is registration only!